### PR TITLE
Triton backend for ONNX Runtime

### PR DESCRIPTION
**Pull request by deadeyegoodwin**

### CODE REVIEW

File: src/backends/backend/onnxruntime/onnxruntime.cc [hunk 1]
```diff
@@ -0,0 +1,1873 @@
...
+//
+class ModelState : public nib::BackendModel {
+ public:
+  static TRITONSERVER_Error* Create(
+      TRITONBACKEND_Model* triton_model, ModelState** state);
+  virtual ~ModelState() = default;
+
+  // Load an ONNX model using 'artifact_name' as the name for the ONNX
+  // file/directory. If 'instance_group_kind' is not
+  // TRITONSERVER_INSTANCEGROUPKIND_AUTO then use it an
```
Comment (GuanLuo) [inline src/backends/backend/onnxruntime/onnxruntime.cc:75]: "use it and"?

File: src/backends/backend/onnxruntime/onnxruntime.cc [hunk 2]
```diff
@@ -0,0 +1,1873 @@
...
+      soptions);
+
+  bool need_lock = false;
+
+  // Execution providers if they are requested...
+  if (instance_group_kind != TRITONSERVER_INSTANCEGROUPKIND_AUTO) {
+    // Default GPU execution provider.
+#ifdef TRITON_ENABLE_GPU
+    if (instance_group_kind == TRITONSERVER_INSTANCEGROUPKIND_GPU) {
+      RETURN_IF_ORT_ERROR(OrtSessionOptionsAppendExecutionProvider_CUDA(
```
Comment (GuanLuo) [inline src/backends/backend/onnxruntime/onnxruntime.cc:220]: The order of calling `OrtSessionOptionsAppendExecutionProvider_XXX()` implies the priority. CUDA should be moved after checking GPU execution accelerator as "default" GPU provider

File: src/backends/backend/onnxruntime/onnxruntime.cc [hunk 3]
```diff
@@ -0,0 +1,1873 @@
...
+                return TRITONSERVER_ErrorNew(
+                    TRITONSERVER_ERROR_INVALID_ARG,
+                    (std::string("unknown Execution Accelerator '") + name +
+                     "' is requested")
+                        .c_str());
+              }
+            }
+          }
+
+          // GPU execution providers
```
Comment (GuanLuo) [inline src/backends/backend/onnxruntime/onnxruntime.cc:270]: CPU

File: src/backends/backend/onnxruntime/onnxruntime.cc [hunk 4]
```diff
@@ -0,0 +1,1873 @@
...
+
+  // ONNX session creation with OpenVINO is not thread-safe,
+  // so multiple creations are serialized with a global lock.
+  static std::mutex global_context_mu;
+  std::unique_lock<std::mutex> glock(global_context_mu, std::defer_lock);
+  if (need_lock) {
+    glock.lock();
+  }
+
+  // Register all op libraries that contain custom operations.
```
Comment (GuanLuo) [inline src/backends/backend/onnxruntime/onnxruntime.cc:313]: Can this part be moved above acquiring the lock?

File: src/backends/backend/onnxruntime/onnxruntime.cc [hunk 5]
```diff
@@ -0,0 +1,1873 @@
...
+    ni::TritonJson::Value model_ops;
+    if (model_config_.Find("model_operations", &model_ops)) {
+      ni::TritonJson::Value op_library_filenames;
+      if (model_ops.Find("op_library_filename", &op_library_filenames)) {
+        for (size_t op_idx = 0; op_idx < op_library_filenames.ArraySize();
+             op_idx++) {
+          std::string op_filename;
+          RETURN_IF_ERROR(
+              op_library_filenames.IndexAsString(op_idx, &op_filename));
+          void* library_handle = nullptr;  // leak this, no harm.
```
Comment (GuanLuo) [inline src/backends/backend/onnxruntime/onnxruntime.cc:324]: Is the comment valid?

File: src/backends/backend/onnxruntime/onnxruntime.cc [hunk 6]
```diff
@@ -0,0 +1,1873 @@
...
+  }
+
+  std::string artifact_name;
+  RETURN_IF_ERROR(
+      ModelConfig().MemberAsString("default_model_filename", &artifact_name));
+
+  std::string model_path;
+  OrtSession* session;
+  OrtAllocator* allocator;
+  // FIXME no early return, must unload session
```
Comment (GuanLuo) [inline src/backends/backend/onnxruntime/onnxruntime.cc:373]: Can use smart pointer to unload on return.

File: src/backends/backend/examples/backend_memory.h [hunk 7]
```diff
@@ -25,44 +25,41 @@
...
-
-  static Status Create(
-      const std::shared_ptr<BackendConfig>& backend_config,
-      std::unique_ptr<OnnxBackendFactory>* factory);
+  // Create a memory allocation using the preferred memory type if
+  // possible. If not possible allocate memory using the next most
+  // appropriate memory type.
+  static TRITONSERVER_Error* Create(
+      const TRITONSERVER_MemoryType preferred_memtype, const size_t byte_size,
+      BackendMemory** mem);
```
[the remark this answers is not in the document]
Comment (deadeyegoodwin) [reply src/backends/backend/examples/backend_memory.h:45]: I want to leave it to the caller to decide that.

File: src/backends/backend/onnxruntime/onnxruntime.cc (hunk 1 above)
Comment (deadeyegoodwin) [reply src/backends/backend/onnxruntime/onnxruntime.cc:75]: fixed

File: src/backends/backend/onnxruntime/onnxruntime.cc (hunk 2 above)
Comment (deadeyegoodwin) [reply src/backends/backend/onnxruntime/onnxruntime.cc:220]: Fixed

File: src/backends/backend/onnxruntime/onnxruntime.cc (hunk 6 above)
Comment (deadeyegoodwin) [reply src/backends/backend/onnxruntime/onnxruntime.cc:373]: Fixed

**Pull request merged**